File: solutions/1599/main.py
from typing import * 

class Solution:
    def minOperationsMaxProfit(self, customers: List[int], boardingCost: int, runningCost: int) -> int:
        calcProf = lambda onboard, i: boardingCost * onboard - runningCost * i

        wait = onboard = max_profit = max_i = 0 
        max_cap = 4 
        for i, c in enumerate(customers):
            wait += c
            # update people number 
            onboard += min(max_cap, wait)
            wait -= min(max_cap, wait)
            # update profit 
            i += 1 
            profit = calcProf(onboard, i)
            if profit > max_profit: 
                max_profit = profit 
                max_i = i
        while wait > 0: 
            onboard += min(max_cap, wait)
            wait -= min(max_cap, wait) 
            i += 1 
            profit = calcProf(onboard, i)
            if profit > max_profit: 
                max_profit = profit 
                max_i = i
        return max_i if max_profit > 0 else -1 
        

if __name__ == "__main__": 
    s = Solution() 

    customers = [8,3]
    boardingCost = 5
    runningCost = 6
    assert s.minOperationsMaxProfit(customers, boardingCost, runningCost) == 3 

File: solutions/1599/main.cpp
class Solution {
public: 
    int minOperationsMaxProfit(vector<int>& customers, int boardingCost, int runningCost) {
        auto calcProf = [boardingCost, runningCost](int i, int onboard) {
            return boardingCost * onboard - runningCost * i;  
        }; 

        int wait, onboard, max_profit, max_cycle; 
        wait = onboard = max_profit = max_cycle = 0; 
        int max_cap = 4; 

        // NOTE, i starts at 1, not at 0 
        for (std::size_t i = 1; i <= customers.size(); ++i) {
            int cur = customers[i-1]; 
            // update the people in queue 
            wait += cur;
            int people = min(max_cap, wait);  
            onboard += people; 
            wait -= people; 

            // update profit 
            int profit = calcProf(i, onboard); 
            if (profit > max_profit) {
                max_profit = profit; 
                max_cycle = i; 
            }
        }

        int j = customers.size(); 
        while (wait > 0) {
            int people = min(max_cap, wait); 
            onboard += people; 
            wait -= people; 

            j += 1; 
            int profit = calcProf(j, onboard); 
            if (profit > max_profit) {
                max_profit = profit; 
                max_cycle = j; 
            }
        }

        if (max_profit > 0) return max_cycle; else return -1; 
    }
};